pkg/noderesourcetopology/least_numa.go

```diff
@@ -0,0 +1,192 @@
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/klog/v2"
+	v1helper "k8s.io/kubernetes/pkg/apis/core/v1/helper"
+	v1qos "k8s.io/kubernetes/pkg/apis/core/v1/helper/qos"
+	"k8s.io/kubernetes/pkg/kubelet/cm/topologymanager/bitmask"
+	"k8s.io/kubernetes/pkg/scheduler/framework"
+	"sigs.k8s.io/scheduler-plugins/pkg/util"
+)
+
+func leastNUMAContainerScopeScore(pod *v1.Pod, zones topologyv1alpha1.ZoneList) (int64, *framework.Status) {
+	nodes := createNUMANodeList(zones)
+	qos := v1qos.GetPodQOS(pod)
+
+	maxNuma := 0
+	// the order how TopologyManager asks for hint is important so doing it in the same order
+	// https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/topologymanager/scope_container.go#L52
+	for _, container := range append(pod.Spec.InitContainers, pod.Spec.Containers...) {
+		identifier := fmt.Sprintf("%s/%s/%s", pod.Namespace, pod.Name, container.Name)
+		numaNodes, res := numaNodesRequired(identifier, qos, nodes, container.Resources.Requests)
+		if res < 1 {
+			errMsg := fmt.Sprintf("cannot calculate NUMA score for container: %s", container.Name)
+			klog.ErrorS(fmt.Errorf(errMsg), "failed to calculate numaNodesRequired for: %s", identifier)
+			return 0, framework.NewStatus(framework.Error, errMsg)
+		}
+
+		if res > maxNuma {
+			maxNuma = res
+		}
+
+		// subtract the resources requested by the container from the given NUMA.
+		// this is necessary, so we won't allocate the same resources for the upcoming containers
+		subtractFromNUMAs(nodes, numaNodes.GetBits(), container.Resources.Requests)
+	}
+
+	return normalizeScore(maxNuma), nil
+}
+
+func leastNUMAPodScopeScore(pod *v1.Pod, zones topologyv1alpha1.ZoneList) (int64, *framework.Status) {
+	nodes := createNUMANodeList(zones)
+	qos := v1qos.GetPodQOS(pod)
+
+	identifier := fmt.Sprintf("%s/%s", pod.Namespace, pod.Name)
+
+	resources := util.GetPodEffectiveRequest(pod)
+
+	_, res := numaNodesRequired(identifier, qos, nodes, resources)
+	if res < 0 {
+		errMsg := fmt.Sprintf("cannot calculate NUMA score for container: %s", identifier)
+		klog.Error(errMsg)
+		return 0, framework.NewStatus(framework.Error, errMsg)
+	}
+
+	return normalizeScore(res), nil
+}
+
+func normalizeScore(numaNodes int) int64 {
+	numaScore := framework.MaxNodeScore / highestNUMAID
+	return framework.MaxNodeScore - int64(numaNodes)*numaScore
+}
+
+func subtractFromNUMAs(numaNodes NUMANodeList, nodes []int, resources v1.ResourceList) {
+	for resName, quantity := range resources {
+		for _, node := range nodes {
+			// quantity is zero no need to iterate through another NUMA node, go to another resource
+			if quantity.IsZero() {
+				break
+			}
+
+			nRes := numaNodes[node].Resources
+			if available, ok := nRes[resName]; ok {
+				switch quantity.Cmp(available) {
+				case 0: // the same
+					// basically zero container resources
+					quantity.Sub(available)
+					// zero NUMA quantity
+					nRes[resName] = resource.MustParse("0")
```

Comment:
nit: I'd move this `MustParse` up and reuse `zeroQty` as needed

Reply:
done